File: packages/python/elizaos/bootstrap/providers/context_bench.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("CONTEXT_BENCH")


async def get_context_bench(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    _ = runtime
    _ = state
    meta = getattr(message, "metadata", None)
    bench_ctx = None
    if meta is not None:
        maybe = getattr(meta, "benchmarkContext", None)
        if isinstance(maybe, str) and maybe.strip():
            bench_ctx = maybe.strip()

    if not bench_ctx:
        return ProviderResult(
            text="",
            values={"benchmark_has_context": False},
            data={},
        )

    return ProviderResult(
        text=f"# Benchmark Context\n{bench_ctx}",
        values={"benchmark_has_context": True},
        data={"benchmarkContext": bench_ctx},
    )


context_bench_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_context_bench,
    position=_spec.get("position"),
    dynamic=_spec.get("dynamic", True),
)

File: packages/rust/src/bootstrap/providers/context_bench.rs
//! Benchmark context provider.

use async_trait::async_trait;

use crate::error::PluginResult;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

pub struct ContextBenchProvider;

#[async_trait]
impl Provider for ContextBenchProvider {
    fn name(&self) -> &'static str {
        "CONTEXT_BENCH"
    }

    fn description(&self) -> &'static str {
        "Benchmark/task context injected by a benchmark harness"
    }

    fn is_dynamic(&self) -> bool {
        true
    }

    async fn get(
        &self,
        _runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let mut bench_ctx: Option<String> = None;
        if let Some(meta) = &message.metadata {
            if let crate::types::memory::MemoryMetadata::Custom(v) = meta {
                if let Some(obj) = v.as_object() {
                    if let Some(s) = obj.get("benchmarkContext").and_then(|x| x.as_str()) {
                        if !s.trim().is_empty() {
                            bench_ctx = Some(s.trim().to_string());
                        }
                    }
                }
            }
        }

        if let Some(ctx) = bench_ctx {
            Ok(ProviderResult {
                text: Some(format!("# Benchmark Context\n{}", ctx)),
                values: Some(
                    [(
                        "benchmark_has_context".to_string(),
                        serde_json::Value::Bool(true),
                    )]
                    .into_iter()
                    .collect(),
                ),
                data: Some(
                    [(
                        "benchmarkContext".to_string(),
                        serde_json::Value::String(ctx),
                    )]
                    .into_iter()
                    .collect(),
                ),
            })
        } else {
            Ok(ProviderResult {
                text: None,
                values: Some(
                    [(
                        "benchmark_has_context".to_string(),
                        serde_json::Value::Bool(false),
                    )]
                    .into_iter()
                    .collect(),
                ),
                data: None,
            })
        }
    }
}
